Add guides to docs (#1522)
hikari-py/hikari: docs/guides/events/advanced.md

# Advanced Usage

```{attention}
This guide only applies to GatewayBot. REST-only applications cannot receive events through the gateway.
```

## wait_for

Sometimes you may want to wait for an event to be received in a procedural manner, then proceed
with execution.

> With `wait_for()` you can block until you receive an event with a given predicate and timeout.

Reviewer: ```suggestion
With `wait_for()` you can block until you receive an event with a given predicate or the function times out.
```

Reviewer 2: Disagree with changing the wording here honestly.

Reviewer: The current wording sounds like the event has a timeout attribute that has to match a certain value, but that's not a thing lol.

Reviewer 3: My suggestion (basically the same as Lunar's):
```suggestion
With `wait_for()` you can block until you receive an event with a given predicate, or until a timeout is reached.
```

In the example below, the bot prompts the user to play a number guessing game.

> Each iteration, the bot waits for the user to input a number, evaluates it and gives an

Reviewer: ```suggestion
Each iteration, the bot waits for the user to input a number, evaluates it then gives an
```

appropiate response.

```py
# We use random for number-generation and asyncio for exception handling
import asyncio
import random

import hikari

> # -- Snip --

Reviewer 3: Considering that this is a guide, and possibly one of the first things new users will read, I don't agree with leaving out bits and replacing them with `# -- Snip --`. At the very least, put this:
```suggestion
bot = hikari.GatewayBot(...)
```

@bot.listen()
async def guessing_game(event: hikari.MessageCreateEvent) -> None:

> 

Reviewer 3: ```suggestion
```

# -- Snip --
```

---

## stream

If you prefer a more functional approach to event handling, you can also use hikari's event streams!

In the example below, we query the user for their 3 most favorite movies and gather them into a list.

```py
@bot.listen()
async def favorite_movie_collector(event: hikari.MessageCreateEvent) -> None:

    if not event.is_human:
        return

    me = bot.get_me()

    # We only want to respond to messages where the bot is pinged
    # Please note that bots by default do not receive message content for messages
    # where they are not pinged or DMd, see the intents section for more information!
    if me.id not in event.message.user_mentions_ids:
        return

    await event.message.respond("Please enter your 3 favorite movies!")

    with bot.stream(hikari.MessageCreateEvent, timeout=None) as stream:
        movies = await (
            stream
            .filter(lambda e: e.author_id == event.author.id and bool(event.message.content))
            .limit(3)
            .map(lambda e: e.message.content)
            .collect(list)
        )

        await event.message.respond(f"Your favorite movies are:```{' '.join(movies)}```")
```

> For more methods available on hikari's `LazyIterator`, check [this page](https://docs.hikari-py.dev/en/latest/reference/hikari/iterators/#hikari.iterators.LazyIterator).

Reviewer 3: ```suggestion
`GatewayBot.stream` returns a `LazyIterator`. For more methods available on hikari's `LazyIterator`, check [this page](https://docs.hikari-py.dev/en/latest/reference/hikari/iterators/#hikari.iterators.LazyIterator).
```
